packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts

```diff
@@ -66,59 +92,60 @@ export const useGridKeyboardNavigation = (
     (params: GridCellParams, event: React.KeyboardEvent) => {
       event.preventDefault();
-
+      const { colIndex } = params;
       const key = mapKey(event);
       const isCtrlPressed = event.ctrlKey || event.metaKey || event.shiftKey;
-
       const cellEl = params.element!;
       cellEl.tabIndex = -1;
-
-      const currentColIndex = Number(cellEl.getAttribute('aria-colindex'));
-      const currentRowIndex = Number(cellEl.getAttribute('data-rowindex'));
+      const rowIndex = Number(cellEl.getAttribute('data-rowindex'));
       let rowCount = totalRowCount;
+
       if (options.pagination && totalRowCount > paginationState.pageSize) {
         rowCount = paginationState.pageSize * (paginationState.page + 1);
       }
 
       let nextCellIndexes: GridCellIndexCoordinates;
       if (isArrowKeys(key)) {
         nextCellIndexes = getNextCellIndexes(key, {
-          colIndex: currentColIndex,
-          rowIndex: currentRowIndex,
+          colIndex,
+          rowIndex,
         });
       } else if (isHomeOrEndKeys(key)) {
         const colIdx = key === 'Home' ? 0 : colCount - 1;
 
         if (!isCtrlPressed) {
           // we go to the current row, first col, or last col!
-          nextCellIndexes = { colIndex: colIdx, rowIndex: currentRowIndex };
+          nextCellIndexes = { colIndex: colIdx, rowIndex };
         } else {
           // In that case we go to first row, first col, or last row last col!
-          let rowIndex = 0;
+          let newRowIndex = 0;
           if (colIdx === 0) {
-            rowIndex = options.pagination ? rowCount - paginationState.pageSize : 0;
+            newRowIndex = options.pagination ? rowCount - paginationState.pageSize : 0;
           } else {
-            rowIndex = rowCount - 1;
+            newRowIndex = rowCount - 1;
           }
-          nextCellIndexes = { colIndex: colIdx, rowIndex };
+          nextCellIndexes = { colIndex: colIdx, rowIndex: newRowIndex };
         }
       } else if (isPageKeys(key) || isSpaceKey(key)) {
         const nextRowIndex =
-          currentRowIndex +
+          rowIndex +
           (key.indexOf('Down') > -1 || isSpaceKey(key)
             ? containerSizes!.viewportPageSize
             : -1 * containerSizes!.viewportPageSize);
-        nextCellIndexes = { colIndex: currentColIndex, rowIndex: nextRowIndex };
+        nextCellIndexes = { colIndex, rowIndex: nextRowIndex };
       } else {
         throw new Error('Material-UI. Key not mapped to navigation behavior.');
       }
 
-      nextCellIndexes.rowIndex = nextCellIndexes.rowIndex <= 0 ? 0 : nextCellIndexes.rowIndex;
+      if (nextCellIndexes.rowIndex < 0) {
+        apiRef.current.setColumnHeaderFocus({ colIndex: nextCellIndexes.colIndex });
+        return;
+      }
+
       nextCellIndexes.rowIndex =
-        nextCellIndexes.rowIndex >= rowCount && rowCount > 0
+        nextCellIndexes.rowIndex !== null && nextCellIndexes.rowIndex >= rowCount && rowCount > 0
```

Comment:
in what case can it be null?

Reply:
That one is not needed, it was a check from before. Removed